exonum-java-binding/core/rust/src/utils/resource_manager/imp.rs

```diff
@@ -160,111 +189,129 @@ mod tests {
-    // state and tests are run concurrently.
-    const MANAGE_HANDLES_FIRST_HANDLE: Handle = 1000;
-    const MANAGE_HANDLES_SECOND_HANDLE: Handle = 2000;
-    const MANAGE_HANDLES_NON_OWNED_HANDLE: Handle = 3000;
-    const DUPLICATED_HANDLE: Handle = 4000;
-    const WRONG_TYPE_HANDLE: Handle = 5000;
-    const WRONG_OWNERSHIP_HANDLE: Handle = 6000;
+    const ZERO_HANDLE: Handle = 0;
 
     #[test]
-    fn manage_handles() {
-        // Add Java-owned handle.
-        enum T1 {}
-        add_handle::<T1>(MANAGE_HANDLES_FIRST_HANDLE);
-        check_handle::<T1>(MANAGE_HANDLES_FIRST_HANDLE);
-        check_handle_impl::<T1>(
-            MANAGE_HANDLES_FIRST_HANDLE,
-            Some(HandleOwnershipType::JavaOwned),
+    fn check_handle_type_and_ownership_ok() {
+        let handle_info = prepare_handle_info();
+        HandlesStorage::check_handle_type_and_ownership::<HandleFooType>(
+            HANDLE_FOO,
+            &handle_info,
+            Some(HandleOwnershipType::NativeOwned),
         );
+    }
 
-        // Add second Java-owned handle.
-        enum T2 {}
-        add_handle::<T2>(MANAGE_HANDLES_SECOND_HANDLE);
-        check_handle::<T2>(MANAGE_HANDLES_SECOND_HANDLE);
-        check_handle_impl::<T2>(
-            MANAGE_HANDLES_SECOND_HANDLE,
-            Some(HandleOwnershipType::JavaOwned),
+    #[test]
+    fn check_handle_type_and_no_ownership_ok() {
+        let handle_info = prepare_handle_info();
+        HandlesStorage::check_handle_type_and_ownership::<HandleFooType>(
+            HANDLE_FOO,
+            &handle_info,
+            None,
         );
+    }
 
-        remove_handle::<T2>(MANAGE_HANDLES_SECOND_HANDLE);
+    #[test]
+    #[should_panic(expected = "Wrong type id for")]
+    fn check_handle_type_wrong() {
+        let handle_info = prepare_handle_info();
+        HandlesStorage::check_handle_type_and_ownership::<HandleBarType>(
+            HANDLE_BAR,
+            &handle_info,
+            None,
+        );
+    }
 
-        // Reuse handle value.
-        add_handle::<T1>(MANAGE_HANDLES_SECOND_HANDLE);
-        check_handle::<T1>(MANAGE_HANDLES_SECOND_HANDLE);
-        check_handle_impl::<T1>(
-            MANAGE_HANDLES_SECOND_HANDLE,
+    #[test]
+    #[should_panic(expected = "handle should be")]
+    fn check_handle_ownership_wrong() {
+        let handle_info = prepare_handle_info();
+        HandlesStorage::check_handle_type_and_ownership::<HandleFooType>(
+            HANDLE_FOO,
+            &handle_info,
             Some(HandleOwnershipType::JavaOwned),
         );
+    }
 
-        // Add native-owned handle.
-        enum T3 {}
-        register_handle::<T3>(MANAGE_HANDLES_NON_OWNED_HANDLE);
-        check_handle::<T3>(MANAGE_HANDLES_NON_OWNED_HANDLE);
-        check_handle_impl::<T3>(
-            MANAGE_HANDLES_NON_OWNED_HANDLE,
-            Some(HandleOwnershipType::NativeOwned),
-        );
+    #[test]
+    fn manage_handles() {
```

Comment:
Please split into more focused tests for operations of rm. There are already some good focused tests below — I'd consider identifying which cases are currently missing and extending the suite of such focused tests.